intro_to_ada.rst

```diff
@@ -1904,23 +1904,747 @@ Abstract data types
+         Generic_Reverse_Array (T => Color, Index => Integer, Array_T => Color_Array);
+
+       My_Colors : Color_Array (1 .. 5) := (Black, Red, Green, Blue, White);
+
+    begin
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+       New_Line;
+       Put_Line ("Reversing My_Color...");
+       New_Line;
+       Reverse_Color_Array (My_Colors);
+
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+    end Test_Reverse_Colors;
+
+As mentioned above, we're abstracting three components of the algorithm:
+
+    - the ``T`` type abstracts the elements of the array
+
+    - the ``Index`` type abstracts the range used for the array
+
+    - the ``Array_T`` type abstracts the array type and makes use of the
+      formal declarations of the ``T`` and ``Index`` types.
+
+
+Abstracting definitions into packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In this section and in the next ones, we will reuse the generic
+reversing algorithm that we discussed in the previous section.  In that
+example, we were declaring three formal types for the
+``Generic_Reverse_Array`` procedure. However, we could abstract the array
+definition into a separate package and reuse it for the generic procedure.
+This could be potentially useful in case we want to create more generic
+procedures for the same array.
+
+In order to achieve this, we start by first specifying a generic package
+that contains the generic array type definition:
+
+.. code-block:: ada
+
+    generic
+       type T is private;
+       type Index is range <>;
+    package Simple_Generic_Array_Pkg is
+       type Array_T is array (Index range <>) of T;
+    end Simple_Generic_Array_Pkg;
+
+As you can see, this definition is the same that we've seen in the
+previous section: we just moved it into a separate package. Now, we have a
+definition of ``Array_T`` that can be reused in multiple places.
+
+The next step is to reuse the ``Simple_Generic_Array_Pkg`` package in the
+``Generic_Reverse_Array`` procedure. By doing this, we can eliminate the
+declaration of the ``Index`` and ``Array_T`` types that we had before,
+since the definition will come from the ``Simple_Generic_Array_Pkg``
+package.
+
+In order to reuse the ``Simple_Generic_Array_Pkg`` package in the
+``Generic_Reverse_Array`` procedure, we need to use a formal package
+declaration in the form:
+
+.. code-block:: ada
+
+    with package P is new Simple_Generic_Array_Pkg(<params>)
+
+This will allow us to reuse definitions from the generic package.
+
+This is the updated version of the our test application for the reversing
+algorithm:
+
+.. code-block:: ada
+
+    with Ada.Text_IO;
+    use  Ada.Text_IO;
+
+    with Simple_Generic_Array_Pkg;
+
+    procedure Test_Reverse_Colors_Simple_Pkg is
+
+       generic
+          type T is private;
+          with package P is new Simple_Generic_Array_Pkg (T => T, Index => Integer);
```

Comment:
It's not a very convincing use of generic formal packages, because you are hard coding the index in a weird place (the reverse algorithm).

I think if you want to keep this example, you should write something like:

~~~ada

generic
   with package P is new Simple_Generic_Array_Pkg (<>)
procedure Reverse_Array (X : in out P.Array_T);
~~~

But it comforts me in the opinion that this belongs in an advanced course :)